# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [17]:
# Imports section
import pandas as pd
import sklearn as sk
import numpy as np

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C','Mols KCL']]
data.drop(columns=['Temperature °C','Mols KCL'])
y = data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=42)
# For grading consistency use random_state=42 

## 3. Perform a Linear Regression

In [13]:
# Use sklearn to train a model on the training set
model = sk.linear_model.LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_datap = pd.DataFrame({'Temperature °C':[469],'Mols KCL': [649]})
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
predicted_output = model.predict(sample_datap)
score = model.score(X_test, y_test)
print("Score of Model: ", score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coef = model.coef_
intercept = model.intercept_
print("Coefficient: " ,coef,"\nIntercept: ", intercept)

Score of Model:  0.8552472077276095
Coefficient:  [ 866.14641337 1032.69506649] 
Intercept:  -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$
h(x) = -409391.47958340764 + 866.14641337 \cdot (\text{Temperature °C}) +
 1032.69506649 \cdot (\text{Mols KCL})
$$

Report on score and explain meaning:
A score of 0.8552472077276095 signifies that roughly 85.5% of the variability in the target varbiable is explained by changes to the two features. Since this number is pretty close to 1 (a perfect score), this is a well made regression model that has a strong correlation.

## 4. Use Cross Validation

In [16]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
scores = sk.model_selection.cross_val_score(model, X, y, cv=5)
print("Cross Validation Scores: ", scores)
# Report on their finding and their significance

Cross Validation Scores:  [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]


Write findings here:
The scores output by the cross validation function show a difference of roughly only +-2, which shows a very tight cluster even across different subsets of data. This low score range indicates that the model is reproducible and robust, further validating the score of 0.855 I recieved earlier.

## 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
# Polynomial regression (degree=2) with k-fold cross-validation
poly = sk.preprocessing.PolynomialFeatures(degree=2)

# Fit only to training data, then transform training and test sets
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)  # <-- fix applied here

# Initialize k-fold cross-validation on polynomial training data
kfold = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
poly_model = sk.linear_model.LinearRegression()

# Cross-validation scores (use polynomial training data here!)
poly_cv_scores = sk.model_selection.cross_val_score(poly_model, X_train_poly, y_train, cv=kfold)
print("Polynomial CV Scores:", poly_cv_scores)

# Train polynomial model fully on training set
poly_model.fit(X_train_poly, y_train)

# Predict using polynomial model on test set to check metrics realistically
y_test_pred = poly_model.predict(X_test_poly)

poly_features = poly.get_feature_names_out(['Temperature °C', 'Mols KCL'])
poly_coef = poly_model.coef_
poly_intercept = poly_model.intercept_
print("Polynomial Coefficients: ", poly_coef)
print("Polynomial Intercept: ", poly_intercept)
print("Polynomial Features: ", poly_features)

# Report on the metrics and output the resultant equation as you did in Part 3.

Polynomial CV Scores: [1. 1. 1. 1. 1.]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Report on the score and interpret:

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [ ]:
# Load the dataset. Then train and evaluate the classification models.

## Results and Conclusion for Classification Experiments

In [ ]:
# Experiments with Neural Network.

## Results and Conclusion for Neural Network Experiments